blocks/monitoring/__init__.py

```diff
@@ -0,0 +1,193 @@
+"""This part of the framework helps you monitor your model during training.
+
+:class:`MonitorChannels` describe quantities that should be monitored. They
+can either be Theano variables (e.g. the objective function, the weight
+norms, etc.) or they can be callable functions (e.g. if you want to use
+NumPy to calculate eigenvalues, or sample text from your language model).
+
+.. todo::
+
+   Monitored Theano variables might depend on inputs beyond those given to
+   the model. There should be a way of providing these to the monitoring
+   channels, collecting them, and compiling them as part of the Theano
+   function that performs monitoring.
+
+"""
+import logging
+from abc import ABCMeta, abstractmethod
+
+from blocks.utils import shared_expression, update_instance
+
+logger = logging.getLogger(__name__)
+
+
+class MonitorChannel(object):
+    """A channel to monitor
+
+    Parameters
+    ----------
+    name : str
+        The name of this monitor channel, should be unique in order to
+        distinguish channels.
+
+    """
+    def __init__(self, name):
+        update_instance(self, locals())
+
+
+class VariableMonitorChannel(MonitorChannel):
```

Comment:
The name is not None, it is whatever theano has set for the name of the 
expression.

I know it is strange to pass |(a, b, a / b), but this is the price for 
having the expresion as a constructor argument, rather than relying on 
it being set once the object os constructed. And the user wll never do 
it himself|

On 1/7/2015 4:19 PM, Bart van Merriënboer wrote:
>
> In blocks/monitoring/__init__.py 
> <https://github.com/bartvm/blocks/pull/90#discussion-diff-22591051>:
>
> > +
> > +class MonitorChannel(object):
> > +    """A channel to monitor
> > +
> > +    Parameters
> > +    ----------
> > +    name : str
> > +        The name of this monitor channel, should be unique in order to
> > +        distinguish channels.
> > +
> > +    """
> > +    def __init__(self, name):
> > +        update_instance(self, locals())
> > +
> > +
> > +class VariableMonitorChannel(MonitorChannel):
>
> Ah, I missed that it was positional argument. Either way, isn't it 
> kind of strange to expect the user to pass |(a, b, a / b)| in the 
> constructor, with the final result still being |name = None| by default?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/bartvm/blocks/pull/90/files#r22591051>.
>